File: android/lib/src/test/kotlin/com/noise/protocol/UnitTests.kt
package com.noise.protocol

import com.noise.protocol.crypto.*
import com.noise.protocol.pattern.HandshakePattern
import com.noise.protocol.pattern.Token
import com.noise.protocol.state.CipherState
import com.noise.protocol.state.HandshakeState
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import kotlin.test.assertContentEquals
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

// MARK: - CipherState Tests

class CipherStateTests {

    @Test
    fun testUnkeyedPassthrough() {
        val cs = CipherState()
        val plaintext = "hello world".toByteArray()
        val result = cs.encryptWithAd(ByteArray(0), plaintext)
        assertContentEquals(plaintext, result, "Unkeyed encrypt should pass through")
    }

    @Test
    fun testKeyedEncryptDecrypt() {
        val key = ByteArray(32) { it.toByte() }
        val cs = CipherState()
        cs.initializeKey(key)
        val ad = "additional data".toByteArray()
        val plaintext = "secret message".toByteArray()
        val ciphertext = cs.encryptWithAd(ad, plaintext)
        assertTrue(ciphertext.size == plaintext.size + 16, "Ciphertext should include 16-byte tag")

        val cs2 = CipherState()
        cs2.initializeKey(key)
        val decrypted = cs2.decryptWithAd(ad, ciphertext)
        assertContentEquals(plaintext, decrypted)
    }

    @Test
    fun testNonceIncrement() {
        val key = ByteArray(32) { it.toByte() }
        val cs = CipherState()
        cs.initializeKey(key)
        assertEquals(0L, cs.getNonce())
        cs.encryptWithAd(ByteArray(0), "test".toByteArray())
        assertEquals(1L, cs.getNonce())
        cs.encryptWithAd(ByteArray(0), "test".toByteArray())
        assertEquals(2L, cs.getNonce())
    }

    @Test
    fun testDecryptionFailureDoesNotIncrementNonce() {
        val key1 = ByteArray(32) { it.toByte() }
        val key2 = ByteArray(32) { (it + 1).toByte() }
        val csEnc = CipherState()
        csEnc.initializeKey(key1)
        val ct = csEnc.encryptWithAd(ByteArray(0), "test".toByteArray())

        val csDec = CipherState()
        csDec.initializeKey(key2) // wrong key
        assertEquals(0L, csDec.getNonce())
        assertThrows<NoiseException.DecryptionFailed> {
            csDec.decryptWithAd(ByteArray(0), ct)
        }
        assertEquals(0L, csDec.getNonce(), "Nonce should not increment on failure")
    }

    @Test
    fun testRekey() {
        val key = ByteArray(32) { it.toByte() }
        val cs1 = CipherState()
        cs1.initializeKey(key)
        val cs2 = CipherState()
        cs2.initializeKey(key)

        cs1.rekey()
        cs2.rekey()

        val plaintext = "after rekey".toByteArray()
        val ct = cs1.encryptWithAd(ByteArray(0), plaintext)
        val pt = cs2.decryptWithAd(ByteArray(0), ct)
        assertContentEquals(plaintext, pt, "Both sides should derive same rekey")
    }
}

// MARK: - Round-trip Handshake Tests

class RoundTripTests {

    @Test
    fun testNNRoundTrip() {
        val initiator = HandshakeState(
            pattern = HandshakePattern.NN,
            initiator = true,
            prologue = ByteArray(0)
        )
        val responder = HandshakeState(
            pattern = HandshakePattern.NN,
            initiator = false,
            prologue = ByteArray(0)
        )

        val (msg1, _) = initiator.writeMessage("hello".toByteArray())
        val (payload1, _) = responder.readMessage(msg1)
        assertContentEquals("hello".toByteArray(), payload1)

        val (msg2, respTransport) = responder.writeMessage("world".toByteArray())
        val (payload2, initTransport) = initiator.readMessage(msg2)
        assertContentEquals("world".toByteArray(), payload2)

        assertNotNull(initTransport)
        assertNotNull(respTransport)

        // Transport phase
        val ct = initTransport!!.sendCipher.encryptWithAd(ByteArray(0), "transport".toByteArray())
        val pt = respTransport!!.receiveCipher.decryptWithAd(ByteArray(0), ct)
        assertContentEquals("transport".toByteArray(), pt)
    }

    @Test
    fun testXXRoundTrip() {
        val initiator = HandshakeState(
            pattern = HandshakePattern.XX,
            initiator = true,
            prologue = ByteArray(0),
            s = NoiseKeyPair.generate()
        )
        val responder = HandshakeState(
            pattern = HandshakePattern.XX,
            initiator = false,
            prologue = ByteArray(0),
            s = NoiseKeyPair.generate()
        )

        val (msg1, _) = initiator.writeMessage("msg1".toByteArray())
        val (p1, _) = responder.readMessage(msg1)
        assertContentEquals("msg1".toByteArray(), p1)

        val (msg2, _) = responder.writeMessage("msg2".toByteArray())
        val (p2, _) = initiator.readMessage(msg2)
        assertContentEquals("msg2".toByteArray(), p2)

        val (msg3, respT) = initiator.writeMessage("msg3".toByteArray())
        val (p3, initT) = responder.readMessage(msg3)
        assertContentEquals("msg3".toByteArray(), p3)

        assertNotNull(initT)
        assertNotNull(respT)

        // Bidirectional transport
        val ct1 = respT!!.sendCipher.encryptWithAd(ByteArray(0), "from init".toByteArray())
        val pt1 = initT!!.receiveCipher.decryptWithAd(ByteArray(0), ct1)
        assertContentEquals("from init".toByteArray(), pt1)

        val ct2 = initT.sendCipher.encryptWithAd(ByteArray(0), "from resp".toByteArray())
        val pt2 = respT.receiveCipher.decryptWithAd(ByteArray(0), ct2)
        assertContentEquals("from resp".toByteArray(), pt2)
    }

    @Test
    fun testIKRoundTrip() {
        val respKp = NoiseKeyPair.generate()
        val initiator = HandshakeState(
            pattern = HandshakePattern.IK,
            initiator = true,
            prologue = ByteArray(0),
            s = NoiseKeyPair.generate(),
            rs = respKp.publicKey
        )
        val responder = HandshakeState(
            pattern = HandshakePattern.IK,
            initiator = false,
            prologue = ByteArray(0),
            s = respKp
        )

        val (msg1, _) = initiator.writeMessage()
        val (_, _) = responder.readMessage(msg1)

        val (msg2, respT) = responder.writeMessage()
        val (_, initT) = initiator.readMessage(msg2)

        assertNotNull(initT)
        assertNotNull(respT)
    }

    @Test
    fun testEmptyPayloadHandshake() {
        val initiator = HandshakeState(
            pattern = HandshakePattern.NN,
            initiator = true,
            prologue = ByteArray(0)
        )
        val responder = HandshakeState(
            pattern = HandshakePattern.NN,
            initiator = false,
            prologue = ByteArray(0)
        )

        val (msg1, _) = initiator.writeMessage()
        val (p1, _) = responder.readMessage(msg1)
        assertEquals(0, p1.size)

        val (msg2, _) = responder.writeMessage()
        val (p2, _) = initiator.readMessage(msg2)
        assertEquals(0, p2.size)
    }
}

// MARK: - DH Tests

class DHTests {

    @Test
    fun testKeyPairGeneration() {
        val kp = NoiseKeyPair.generate()
        assertEquals(32, kp.publicKey.size)
    }

    @Test
    fun testDeterministicKeyPair() {
        val privKey = "893e28b9dc6ca8d611ab664754b8ceb7bac5117349a4439a6b0569da977c464a".hexToBytes()
        val kp = NoiseKeyPair.fromPrivateKey(privKey)
        val expectedPub = "ca35def5ae56cec33dc2036731ab14896bc4c75dbb07a61f879f8e3afa4c7944".hexToBytes()
        assertContentEquals(expectedPub, kp.publicKey)
    }

    @Test
    fun testDHAgreement() {
        val kp1 = NoiseKeyPair.generate()
        val kp2 = NoiseKeyPair.generate()
        val shared1 = kp1.dh(kp2.publicKey)
        val shared2 = kp2.dh(kp1.publicKey)
        assertContentEquals(shared1, shared2, "DH should be commutative")
        assertEquals(32, shared1.size)
    }
}

// MARK: - Error Handling Tests

class ErrorHandlingTests {

    @Test
    fun testWriteAfterComplete() {
        val initiator = HandshakeState(
            pattern = HandshakePattern.NN,
            initiator = true,
            prologue = ByteArray(0)
        )
        val responder = HandshakeState(
            pattern = HandshakePattern.NN,
            initiator = false,
            prologue = ByteArray(0)
        )

        val (msg1, _) = initiator.writeMessage()
        responder.readMessage(msg1)
        val (msg2, _) = responder.writeMessage()
        initiator.readMessage(msg2)

        assertThrows<NoiseException.HandshakeAlreadyComplete> {
            initiator.writeMessage()
        }
    }

    @Test
    fun testReadOnWriteTurn() {
        val initiator = HandshakeState(
            pattern = HandshakePattern.NN,
            initiator = true,
            prologue = ByteArray(0)
        )
        assertThrows<NoiseException> {
            initiator.readMessage(ByteArray(32))
        }
    }

    @Test
    fun testDecryptionFailure() {
        val key = ByteArray(32) { it.toByte() }
        val cs = CipherState()
        cs.initializeKey(key)
        assertThrows<NoiseException.DecryptionFailed> {
            cs.decryptWithAd(ByteArray(0), ByteArray(20))
        }
    }

    @Test
    fun testUnknownPattern() {
        assertThrows<NoiseException.UnknownPattern> {
            HandshakePattern.named("NotAPattern")
        }
    }
}

// MARK: - Pattern Definition Tests

class PatternDefinitionTests {

    @Test
    fun testNNDefinition() {
        val p = HandshakePattern.NN
        assertEquals(2, p.messagePatterns.size)
        assertEquals(listOf(Token.E), p.messagePatterns[0])
        assertEquals(listOf(Token.E, Token.EE), p.messagePatterns[1])
        assertTrue(p.initiatorPreMessage.isEmpty())
        assertTrue(p.responderPreMessage.isEmpty())
    }

    @Test
    fun testXXDefinition() {
        val p = HandshakePattern.XX
        assertEquals(3, p.messagePatterns.size)
        assertEquals(listOf(Token.E), p.messagePatterns[0])
        assertEquals(listOf(Token.E, Token.EE, Token.S, Token.ES), p.messagePatterns[1])
        assertEquals(listOf(Token.S, Token.SE), p.messagePatterns[2])
    }

    @Test
    fun testIKDefinition() {
        val p = HandshakePattern.IK
        assertEquals(2, p.messagePatterns.size)
        assertEquals(listOf(Token.E, Token.ES, Token.S, Token.SS), p.messagePatterns[0])
        assertEquals(listOf(Token.E, Token.EE, Token.SE), p.messagePatterns[1])
        assertEquals(listOf(Token.S), p.responderPreMessage)
    }

    @Test
    fun testNKpsk0Definition() {
        val p = HandshakePattern.NKpsk0
        assertEquals(listOf(Token.PSK, Token.E, Token.ES), p.messagePatterns[0])
        assertEquals(listOf(Token.E, Token.EE), p.messagePatterns[1])
        assertEquals(listOf(Token.S), p.responderPreMessage)
    }

    @Test
    fun testIKpsk2Definition() {
        val p = HandshakePattern.IKpsk2
        assertEquals(listOf(Token.E, Token.ES, Token.S, Token.SS), p.messagePatterns[0])
        assertEquals(listOf(Token.E, Token.EE, Token.SE, Token.PSK), p.messagePatterns[1])
        assertEquals(listOf(Token.S), p.responderPreMessage)
    }

    @Test
    fun testXXfallbackDefinition() {
        val p = HandshakePattern.XXfallback
        assertEquals(2, p.messagePatterns.size)
        assertEquals(listOf(Token.E), p.responderPreMessage)
    }

    @Test
    fun testNamedPatternLookup() {
        assertNotNull(HandshakePattern.named("NN"))
        assertNotNull(HandshakePattern.named("XX"))
        assertNotNull(HandshakePattern.named("IK"))
        assertNotNull(HandshakePattern.named("XXfallback"))
        assertThrows<NoiseException.UnknownPattern> {
            HandshakePattern.named("ZZ")
        }
    }

    @Test
    fun testAllPatternsExist() {
        val expectedPatterns = listOf(
            "N", "K", "X",
            "NN", "NK", "NX", "XN", "XK", "XX", "KN", "KK", "KX", "IN", "IK", "IX",
            "XXfallback", "NKpsk0", "IKpsk2"
        )
        for (name in expectedPatterns) {
            assertNotNull(HandshakePattern.all[name], "Pattern $name should exist")
        }
    }
}

// MARK: - Channel Binding Tests

class ChannelBindingTests {

    @Test
    fun testHandshakeHashMatches() {
        val initKp = NoiseKeyPair.generate()
        val respKp = NoiseKeyPair.generate()

        val initiator = HandshakeState(
            pattern = HandshakePattern.XX,
            initiator = true,
            prologue = ByteArray(0),
            s = initKp
        )
        val responder = HandshakeState(
            pattern = HandshakePattern.XX,
            initiator = false,
            prologue = ByteArray(0),
            s = respKp
        )

        val (m1, _) = initiator.writeMessage()
        responder.readMessage(m1)
        val (m2, _) = responder.writeMessage()
        initiator.readMessage(m2)
        val (m3, respT) = initiator.writeMessage()
        val (_, initT) = responder.readMessage(m3)

        assertNotNull(initT)
        assertNotNull(respT)
        assertContentEquals(
            initT!!.handshakeHash,
            respT!!.handshakeHash,
            "Both sides should have same handshake hash"
        )
    }

    @Test
    fun testRemoteStaticKeyExchange() {
        val initKp = NoiseKeyPair.generate()
        val respKp = NoiseKeyPair.generate()

        val initiator = HandshakeState(
            pattern = HandshakePattern.XX,
            initiator = true,
            prologue = ByteArray(0),
            s = initKp
        )
        val responder = HandshakeState(
            pattern = HandshakePattern.XX,
            initiator = false,
            prologue = ByteArray(0),
            s = respKp
        )

        val (m1, _) = initiator.writeMessage()
        responder.readMessage(m1)
        val (m2, _) = responder.writeMessage()
        initiator.readMessage(m2)
        val (m3, respT) = initiator.writeMessage()
        val (_, initT) = responder.readMessage(m3)

        assertNotNull(initT)
        assertNotNull(respT)
        assertContentEquals(respKp.publicKey, respT!!.remoteStaticKey, "Initiator should know responder's static key")
        assertContentEquals(initKp.publicKey, initT!!.remoteStaticKey, "Responder should know initiator's static key")
    }
}

// MARK: - Crypto Primitives Tests

class CryptoPrimitivesTests {

    @Test
    fun testSHA256KnownVector() {
        val input = "abc".toByteArray()
        val expected = "ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad".hexToBytes()
        assertContentEquals(expected, NoiseHash.hash(input))
    }

    @Test
    fun testHKDF2Outputs() {
        val ck = ByteArray(32) { it.toByte() }
        val ikm = ByteArray(32) { (it + 32).toByte() }
        val outputs = NoiseHash.hkdf(ck, ikm, 2)
        assertEquals(2, outputs.size)
        assertEquals(32, outputs[0].size)
        assertEquals(32, outputs[1].size)
    }

    @Test
    fun testHKDF3Outputs() {
        val ck = ByteArray(32) { it.toByte() }
        val ikm = ByteArray(32) { (it + 32).toByte() }
        val outputs = NoiseHash.hkdf(ck, ikm, 3)
        assertEquals(3, outputs.size)
        for (o in outputs) assertEquals(32, o.size)
    }

    @Test
    fun testChaCha20Poly1305RoundTrip() {
        val key = ByteArray(32) { it.toByte() }
        val nonce = 42L
        val ad = "test".toByteArray()
        val plaintext = "hello world!".toByteArray()
        val ct = NoiseCipher.encrypt(key, nonce, ad, plaintext)
        assertEquals(plaintext.size + 16, ct.size)
        val pt = NoiseCipher.decrypt(key, nonce, ad, ct)
        assertContentEquals(plaintext, pt)
    }
}

File: ios/Tests/NoiseProtocolTests/UnitTests.swift
import XCTest
@testable import NoiseProtocol
import Foundation

// MARK: - CipherState Tests

final class CipherStateTests: XCTestCase {

    func testNoKeyPassthrough() throws {
        let cs = CipherState()
        let plaintext = Data("hello".utf8)
        let ad = Data("ad".utf8)
        let result = try cs.encryptWithAd(ad, plaintext: plaintext)
        XCTAssertEqual(result, plaintext, "Without key, encryptWithAd should return plaintext")

        let decrypted = try cs.decryptWithAd(ad, ciphertext: plaintext)
        XCTAssertEqual(decrypted, plaintext, "Without key, decryptWithAd should return ciphertext")
    }

    func testEncryptDecryptRoundTrip() throws {
        let cs = CipherState()
        let key = Data(repeating: 0x42, count: 32)
        cs.initializeKey(key)
        XCTAssertTrue(cs.hasKey)

        let ad = Data("associated data".utf8)
        let plaintext = Data("secret message".utf8)

        let ciphertext = try cs.encryptWithAd(ad, plaintext: plaintext)
        XCTAssertNotEqual(ciphertext, plaintext)
        XCTAssertEqual(ciphertext.count, plaintext.count + 16) // 16-byte AEAD tag

        // New CipherState for decryption (same key, nonce resets to 0)
        let cs2 = CipherState()
        cs2.initializeKey(key)
        let decrypted = try cs2.decryptWithAd(ad, ciphertext: ciphertext)
        XCTAssertEqual(decrypted, plaintext)
    }

    func testNonceIncrement() throws {
        let cs = CipherState()
        cs.initializeKey(Data(repeating: 0x01, count: 32))
        XCTAssertEqual(cs.getNonce(), 0)

        _ = try cs.encryptWithAd(Data(), plaintext: Data("a".utf8))
        XCTAssertEqual(cs.getNonce(), 1)

        _ = try cs.encryptWithAd(Data(), plaintext: Data("b".utf8))
        XCTAssertEqual(cs.getNonce(), 2)
    }

    func testDecryptionFailureDoesNotIncrementNonce() throws {
        let cs = CipherState()
        cs.initializeKey(Data(repeating: 0x01, count: 32))

        let badCiphertext = Data(repeating: 0xFF, count: 32)
        XCTAssertThrowsError(try cs.decryptWithAd(Data(), ciphertext: badCiphertext))
        XCTAssertEqual(cs.getNonce(), 0, "Nonce should not increment after decryption failure")
    }

    func testRekey() throws {
        let cs = CipherState()
        let key = Data(repeating: 0xAB, count: 32)
        cs.initializeKey(key)

        let plaintext = Data("test".utf8)
        let ct1 = try cs.encryptWithAd(Data(), plaintext: plaintext)

        // Rekey and encrypt again — should produce different ciphertext
        try cs.rekey()
        cs.setNonce(0)
        let ct2 = try cs.encryptWithAd(Data(), plaintext: plaintext)
        XCTAssertNotEqual(ct1, ct2, "Rekey should change the cipher key")
    }

    func testRekeyWithoutKeyThrows() {
        let cs = CipherState()
        XCTAssertThrowsError(try cs.rekey())
    }
}

// MARK: - Handshake Round-Trip Tests (random keys)

final class HandshakeRoundTripTests: XCTestCase {

    func testNNRoundTrip() throws {
        let initiator = HandshakeState(pattern: .NN, initiator: true)
        let responder = HandshakeState(pattern: .NN, initiator: false)

        // Message 1: initiator → responder
        let (msg1, _) = try initiator.writeMessage(payload: Data("hello".utf8))
        let (payload1, _) = try responder.readMessage(msg1)
        XCTAssertEqual(String(data: payload1, encoding: .utf8), "hello")

        // Message 2: responder → initiator
        let (msg2, respTransport) = try responder.writeMessage(payload: Data("world".utf8))
        let (payload2, initTransport) = try initiator.readMessage(msg2)
        XCTAssertEqual(String(data: payload2, encoding: .utf8), "world")

        XCTAssertNotNil(initTransport)
        XCTAssertNotNil(respTransport)

        // Transport round-trip
        let secret = Data("transport data".utf8)
        let ct = try initTransport!.sendCipher.encryptWithAd(Data(), plaintext: secret)
        let pt = try respTransport!.receiveCipher.decryptWithAd(Data(), ciphertext: ct)
        XCTAssertEqual(pt, secret)
    }

    func testXXRoundTrip() throws {
        let initS = NoiseKeyPair()
        let respS = NoiseKeyPair()

        let initiator = HandshakeState(
            pattern: .XX, initiator: true,
            s: initS
        )
        let responder = HandshakeState(
            pattern: .XX, initiator: false,
            s: respS
        )

        // Message 1: initiator → responder
        let (msg1, _) = try initiator.writeMessage()
        let (_, _) = try responder.readMessage(msg1)

        // Message 2: responder → initiator
        let (msg2, _) = try responder.writeMessage()
        let (_, _) = try initiator.readMessage(msg2)

        // Message 3: initiator → responder
        let (msg3, initT) = try initiator.writeMessage(payload: Data("done".utf8))
        let (p3, respT) = try responder.readMessage(msg3)

        XCTAssertEqual(String(data: p3, encoding: .utf8), "done")
        XCTAssertNotNil(initT)
        XCTAssertNotNil(respT)

        // Both sides should agree on handshake hash
        XCTAssertEqual(initT!.handshakeHash, respT!.handshakeHash)

        // Both sides should know each other's static keys
        XCTAssertEqual(initT!.remoteStaticKey, respS.publicKey)
        XCTAssertEqual(respT!.remoteStaticKey, initS.publicKey)
    }

    func testIKRoundTrip() throws {
        let initS = NoiseKeyPair()
        let respS = NoiseKeyPair()

        let initiator = HandshakeState(
            pattern: .IK, initiator: true,
            s: initS,
            rs: respS.publicKey
        )
        let responder = HandshakeState(
            pattern: .IK, initiator: false,
            s: respS,
            rs: initS.publicKey
        )

        let (msg1, _) = try initiator.writeMessage(payload: Data("ik-hello".utf8))
        let (p1, _) = try responder.readMessage(msg1)
        XCTAssertEqual(String(data: p1, encoding: .utf8), "ik-hello")

        let (msg2, respT) = try responder.writeMessage()
        let (_, initT) = try initiator.readMessage(msg2)

        XCTAssertNotNil(initT)
        XCTAssertNotNil(respT)
        XCTAssertEqual(initT!.handshakeHash, respT!.handshakeHash)
    }

    func testNKRoundTrip() throws {
        let respS = NoiseKeyPair()

        let initiator = HandshakeState(
            pattern: .NK, initiator: true,
            rs: respS.publicKey
        )
        let responder = HandshakeState(
            pattern: .NK, initiator: false,
            s: respS
        )

        let (msg1, _) = try initiator.writeMessage()
        let (_, _) = try responder.readMessage(msg1)

        let (msg2, respT) = try responder.writeMessage()
        let (_, initT) = try initiator.readMessage(msg2)

        XCTAssertNotNil(initT)
        XCTAssertNotNil(respT)
        XCTAssertEqual(initT!.handshakeHash, respT!.handshakeHash)
    }

    func testEmptyPayloads() throws {
        let initiator = HandshakeState(pattern: .NN, initiator: true)
        let responder = HandshakeState(pattern: .NN, initiator: false)

        let (msg1, _) = try initiator.writeMessage()
        let (p1, _) = try responder.readMessage(msg1)
        XCTAssertTrue(p1.isEmpty)

        let (msg2, _) = try responder.writeMessage()
        let (p2, _) = try initiator.readMessage(msg2)
        XCTAssertTrue(p2.isEmpty)
    }
}

// MARK: - Error Handling Tests

final class ErrorHandlingTests: XCTestCase {

    func testHandshakeAlreadyComplete() throws {
        let initiator = HandshakeState(pattern: .NN, initiator: true)
        let responder = HandshakeState(pattern: .NN, initiator: false)

        let (msg1, _) = try initiator.writeMessage()
        let (_, _) = try responder.readMessage(msg1)
        let (msg2, _) = try responder.writeMessage()
        let (_, _) = try initiator.readMessage(msg2)

        // Should throw on further writes
        XCTAssertThrowsError(try initiator.writeMessage()) { error in
            guard case NoiseError.handshakeAlreadyComplete = error else {
                XCTFail("Expected handshakeAlreadyComplete, got \(error)")
                return
            }
        }
    }

    func testDecryptionFailureWithWrongKey() throws {
        let cs = CipherState()
        cs.initializeKey(Data(repeating: 0x01, count: 32))
        let ct = try cs.encryptWithAd(Data(), plaintext: Data("test".utf8))

        let cs2 = CipherState()
        cs2.initializeKey(Data(repeating: 0x02, count: 32))
        XCTAssertThrowsError(try cs2.decryptWithAd(Data(), ciphertext: ct)) { error in
            guard case NoiseError.decryptionFailed = error else {
                XCTFail("Expected decryptionFailed, got \(error)")
                return
            }
        }
    }

    func testWrongADDecryptionFails() throws {
        let key = Data(repeating: 0x42, count: 32)
        let cs1 = CipherState()
        cs1.initializeKey(key)
        let ct = try cs1.encryptWithAd(Data("ad1".utf8), plaintext: Data("test".utf8))

        let cs2 = CipherState()
        cs2.initializeKey(key)
        XCTAssertThrowsError(try cs2.decryptWithAd(Data("wrong-ad".utf8), ciphertext: ct))
    }
}

// MARK: - DH Tests

final class DHTests: XCTestCase {

    func testKeyPairGeneration() {
        let kp = NoiseKeyPair()
        XCTAssertEqual(kp.publicKey.count, DHLEN)
    }

    func testDeterministicKeyPair() throws {
        let priv = Data(hex: "893e28b9dc6ca8d611ab664754b8ceb7bac5117349a4439a6b0569da977c464a")
        let kp = try NoiseKeyPair(privateKeyData: priv)
        let expectedPub = Data(hex: "ca35def5ae56cec33dc2036731ab14896bc4c75dbb07a61f879f8e3afa4c7944")
        XCTAssertEqual(kp.publicKey.hex, expectedPub.hex)
    }

    func testDHSymmetry() throws {
        let alice = NoiseKeyPair()
        let bob = NoiseKeyPair()

        let shared1 = try alice.dh(remotePublicKey: bob.publicKey)
        let shared2 = try bob.dh(remotePublicKey: alice.publicKey)
        XCTAssertEqual(shared1, shared2, "DH should be symmetric")
    }

    func testDHOutputLength() throws {
        let alice = NoiseKeyPair()
        let bob = NoiseKeyPair()
        let shared = try alice.dh(remotePublicKey: bob.publicKey)
        XCTAssertEqual(shared.count, DHLEN)
    }
}

// MARK: - Pattern Tests

final class PatternTests: XCTestCase {

    func testPatternLookup() {
        XCTAssertNotNil(HandshakePattern.all["NN"])
        XCTAssertNotNil(HandshakePattern.all["NK"])
        XCTAssertNotNil(HandshakePattern.all["XX"])
        XCTAssertNotNil(HandshakePattern.all["IK"])
        XCTAssertNotNil(HandshakePattern.all["XXfallback"])
        XCTAssertNotNil(HandshakePattern.all["NKpsk0"])
        XCTAssertNotNil(HandshakePattern.all["IKpsk2"])
    }

    func testOneWayPatterns() {
        XCTAssertNotNil(HandshakePattern.all["N"])
        XCTAssertNotNil(HandshakePattern.all["K"])
        XCTAssertNotNil(HandshakePattern.all["X"])
    }

    func testDeferredPatterns() {
        let deferred = ["NK1", "NX1", "X1N", "X1K", "XK1", "X1K1", "X1X", "XX1", "X1X1",
                        "K1N", "K1K", "KK1", "K1K1", "K1X", "KX1", "K1X1",
                        "I1N", "I1K", "IK1", "I1K1", "I1X", "IX1", "I1X1"]
        for name in deferred {
            XCTAssertNotNil(HandshakePattern.all[name], "Missing deferred pattern: \(name)")
        }
    }

    func testWithPSKModifier() {
        let nk = HandshakePattern.NK
        let nkpsk0 = nk.withPSK(positions: [0])
        XCTAssertEqual(nkpsk0.name, "NKpsk0")
        XCTAssertEqual(nkpsk0.messagePatterns[0].first, .psk)
    }

    func testXXfallbackDefinition() {
        let p = HandshakePattern.XXfallback
        XCTAssertEqual(p.name, "XXfallback")
        XCTAssertTrue(p.responderPreMessage.contains(.e))
        XCTAssertTrue(p.initiatorPreMessage.isEmpty)
        XCTAssertEqual(p.messagePatterns.count, 2)
    }

    func testNKpsk0Definition() {
        let p = HandshakePattern.NKpsk0
        XCTAssertEqual(p.messagePatterns[0], [.psk, .e, .es])
        XCTAssertEqual(p.messagePatterns[1], [.e, .ee])
        XCTAssertEqual(p.responderPreMessage, [.s])
    }

    func testIKpsk2Definition() {
        let p = HandshakePattern.IKpsk2
        XCTAssertEqual(p.messagePatterns[0], [.e, .es, .s, .ss])
        XCTAssertEqual(p.messagePatterns[1], [.e, .ee, .se, .psk])
        XCTAssertEqual(p.responderPreMessage, [.s])
    }
}

// MARK: - Channel Binding Tests

final class ChannelBindingTests: XCTestCase {

    func testHandshakeHashMatch() throws {
        let initS = NoiseKeyPair()
        let respS = NoiseKeyPair()

        let initiator = HandshakeState(
            pattern: .XX, initiator: true,
            prologue: Data("test-prologue".utf8),
            s: initS
        )
        let responder = HandshakeState(
            pattern: .XX, initiator: false,
            prologue: Data("test-prologue".utf8),
            s: respS
        )

        let (m1, _) = try initiator.writeMessage()
        let (_, _) = try responder.readMessage(m1)
        let (m2, _) = try responder.writeMessage()
        let (_, _) = try initiator.readMessage(m2)
        let (m3, iT) = try initiator.writeMessage()
        let (_, rT) = try responder.readMessage(m3)

        XCTAssertEqual(iT!.handshakeHash, rT!.handshakeHash)
        // Hash should be 32 bytes
        XCTAssertEqual(iT!.handshakeHash.count, 32)
    }

    func testDifferentPrologueProducesDifferentHash() throws {
        func doHandshake(prologue: Data) throws -> Data {
            let initiator = HandshakeState(pattern: .NN, initiator: true, prologue: prologue)
            let responder = HandshakeState(pattern: .NN, initiator: false, prologue: prologue)
            let (m1, _) = try initiator.writeMessage()
            let (_, _) = try responder.readMessage(m1)
            let (m2, _) = try responder.writeMessage()
            let (_, t) = try initiator.readMessage(m2)
            return t!.handshakeHash
        }

        let h1 = try doHandshake(prologue: Data("prologue-A".utf8))
        let h2 = try doHandshake(prologue: Data("prologue-B".utf8))
        XCTAssertNotEqual(h1, h2)
    }
}
